sql: add full table or index scan count metric

barryhe2000 opened the pull request:
Previously, this metric was not available, but is now added so that users can
see when full table scans or full index scans are being used for their queries.

Fixes: #58653

Release note (ui change): User can see time series custom chart in admin ui
 for full table or index scans.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59261)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice work! I have some nits, but the code looks good to me in general.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-:-MRam4HD0uHuEGvkOTCb:b-of1p0b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @barryhe2000)

___
*[pkg/sql/conn_executor.go, line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-MRakGzW0D6DWnSCMwB4:-MRakGzW0D6DWnSCMwB5:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b979fd6a5216b09c0dc8a17a3722ec3343fb4c3/pkg/sql/conn_executor.go#L298)):*
> ```Go
> 
> func makeMetrics(internal bool) Metrics {
> 	// fullTableIndexScan is only instantiated for non-internal queries
> ```

nit: missing period.
___
*[pkg/sql/conn_executor_exec.go, line 917 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-MRakXUE0ZULKGMOF14U:-MRakXUE0ZULKGMOF14V:b-v9t6ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b979fd6a5216b09c0dc8a17a3722ec3343fb4c3/pkg/sql/conn_executor_exec.go#L917)):*
> ```Go
> 	flags := planner.curPlan.flags
> 
> 	// We don't execute the statement if:
> ```

nit: this comment should be moved below, to right before returning an error that stops the statement execution.
___
*[pkg/sql/exec_util.go, line 478 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-MRaktcZ-nFXdazTneYX:-MRaktcZ-nFXdazTneYY:b7p216m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b979fd6a5216b09c0dc8a17a3722ec3343fb4c3/pkg/sql/exec_util.go#L478)):*
> ```Go
> 	}
> 	MetaFullTableOrIndexScan = metric.Metadata{
> 		Name:        "sql.full.table.index.scan.count",
> ```

nit: I think I'd use `sql.full_index_scan.count` ("full table" scan is a "full primary index" scan) or just `sql.full_scan.count`. Maybe @jordanlewis or @asubiotto have a better suggestion.
___
*[pkg/sql/executor_statement_metrics.go, line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-MRalDNl2fj1ukAjByJh:-MRalDNl2fj1ukAjByJi:b-5o753w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b979fd6a5216b09c0dc8a17a3722ec3343fb4c3/pkg/sql/executor_statement_metrics.go#L143)):*
> ```Go
> 	FailureCount *metric.Counter
> 
> 	// FullTableOrIndexScanCount counts the number of full table or index scans.
> ```

nit: I'd mention that it is only for user-initiated queries.


<!-- Sent from Reviewable.io -->

barryhe2000 commented:
Fixed the nits that you mentioned @yuzefovich

asubiotto commented:
Nice! Did you manage to find a testing strategy for metrics?

awoods187 commented:
Mind posting a sample output for a workload that we know about--say TPC-C?

yuzefovich commented:
Do we wanna add a dashboard for this metric? I initially thought it wasn't necessary since I assumed that the full table/index scan info is useful only during the debugging, but I'm not sure whether that assumption is correct - it might make sense to have a dashboard for it at the very bottom of `SQL` dashboard. Thoughts?

asubiotto commented:
I'm not convinced it's useful as a `SQL` dashboard level metric. If everything is well, that graph should flatline at zero, which isn't very interesting. It can currently be accessed through the debug page

asubiotto commented:
I think what @awoods187 is saying though, is if we could see the output of this graph (regardless of where it ends up being) for a sample workload.

barryhe2000 commented:
<img width="1302" alt="Screen Shot 2021-01-20 at 8 23 45 PM" src="https://user-images.githubusercontent.com/49596388/105736248-71871080-5ee9-11eb-841b-1dffc14cb8b0.png">

Metric currently looks something like this

awoods187 commented:
I actually think we should add this as a chart in the DB Console. Operators often want some measurement of how many scans are occurring at the cluster level--in many cases full table scans are the best option so I wouldn't expect this metric to be zero. I like the way it looks in the custom chart and think we should keep this metric at the bottom of SQL reports.

asubiotto commented:
OK. In that case, @barryhe2000 note that you probably want to use a non-negative rate option so the timeseries doesn't show the count since the beginning of time, rather the count in a given time interval.

yuzefovich reviewed (commented):
There are now 4 commits in this PR, we should clean that up so that only one commit remains.

> Reviewed 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-:-MS534fvBDxdOuVq2nUO:bx1suwp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @barryhe2000)

___
*[pkg/sql/conn_executor.go, line 904 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-MS52O1kBvazgz6w7QW-:-MS52O1kBvazgz6w7QW0:bvp45uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5456b6f5aaf6b21767f81076c799c5fe3346d631/pkg/sql/conn_executor.go#L904)):*
> ```Go
> 
> 	// distSQLMetrics contains distributed metrics to be accounted for.
> 	distSQLMetrics *execinfra.DistSQLMetrics
> ```

This is no longer needed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59261#-:-MS55spMCy6am8a5g3VD:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @barryhe2000)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59261)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

barryhe2000 committed:
sql: add full table or index scan count to metrics

This new metric allows for users to see a time series of their full table
or index scans in the advanced debug console. This metric is part of
EngineMetrics, so there's a corresponding internal metric that counts
internal full table or index scans from internal engine queries.

Release note (ui change): User can see time series of full table or index
scans in advanced debug console.

barryhe2000 commented:
bors r+

yuzefovich commented:
bors r-

Hm, I'm guessing the bors is complaining about the license/cla check.

Also, I see that you added the chart - I think we discussed putting it at the very bottom of the SQL dashboard, and currently it's in the middle.

barryhe2000 commented:
CI was failing a test because the chart wasn't there, I can move it. Is there anything doable about the license/cla check? Not sure why it isn't passing

barryhe2000 commented:
bors r+
